<a href="https://colab.research.google.com/github/ChelseaLiu0822/LDRPM-Loan-Default-Risk-Prediction-Model/blob/main/notebooks/DataPreprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Objectives:
- use dask to load data
- explore data
- transform data
- aggregate data
- merge data
- output a tiny parquet file

## refs:   
eda: https://www.kaggle.com/code/taichiuemura/home-credit-eda  

feature group: https://www.kaggle.com/code/seoyunje/home-credit-feature-group/notebook

If you want to know data description, check out below

https://www.kaggle.com/competitions/home-credit-credit-risk-model-stability/discussion/473950

If you want to know feature column description, check out below

https://www.kaggle.com/datasets/kononenko/home-credit-enhanced-feature-definitions  
How to aggregate dataframes:  
https://www.kaggle.com/code/greysky/home-credit-baseline  
A more detailed description file for features: https://www.kaggle.com/datasets/kononenko/home-credit-enhanced-feature-definitions  
About num_group:
1. https://www.kaggle.com/competitions/home-credit-credit-risk-model-stability/discussion/476763  
2. https://www.kaggle.com/competitions/home-credit-credit-risk-model-stability/discussion/475548



# Get Original Data  
ref link: https://stackoverflow.com/questions/600079/how-do-i-clone-a-subdirectory-only-of-a-git-repository/28039894#28039894

## connect to the github(Optional)
please check [getOriginalData.ipynb](https://colab.research.google.com/drive/1C0-ILD3LMxTsIZfsyuNCyi_d_i0hk_Jb) to see "Fail to clone: fatal: could not read Username for 'https://github.com': No such device or address"

In [ ]:
!mkdir -p /root/.ssh

In [ ]:
!ssh-keyscan -t rsa github.com >> /root/.ssh/known_hosts

# github.com:22 SSH-2.0-babeld-05989c77


In [ ]:
!ssh-keygen -t rsa -C "ql2547@nyu.edu"
# !ssh-keygen -t rsa -C "ys5795@nyu.edu"
# !ssh-keygen -t rsa -C "zg2442@nyu.edu"
# input your email

Generating public/private rsa key pair.
Enter file in which to save the key (/root/.ssh/id_rsa): 
/root/.ssh/id_rsa already exists.
Overwrite (y/n)? y
Enter passphrase (empty for no passphrase): y
Enter same passphrase again: 
Passphrases do not match.  Try again.
Enter passphrase (empty for no passphrase): 
Enter same passphrase again: 
Your identification has been saved in /root/.ssh/id_rsa
Your public key has been saved in /root/.ssh/id_rsa.pub
The key fingerprint is:
SHA256:bfPTwLWSi8vNgeqtXjrJv55LdxkY5cy7DRaCFKCfqCU ql2547@nyu.edu
The key's randomart image is:
+---[RSA 3072]----+
|        ..o.  .  |
|       . . . =   |
|      .   . o *  |
|       o o . * + |
|    E o S + * *  |
|     +   . = * * |
|    .  . .= * = .|
|        +B * +   |
|       o*=Xoo    |
+----[SHA256]-----+


In [ ]:
!ls /root/.ssh

id_rsa	id_rsa.pub  known_hosts  known_hosts.old


In [ ]:
!cat /root/.ssh/id_rsa.pub

ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQDN1M1aA/WztOzXCIaGHBxi7xwoxMbN4/jvcUN0EdUVxqSuJFe8tTDnhdoV6ftHq+add/pHCFqn6yo+ZOIvlkRpDsGQgmDtzMGtDxc2A7Jak2w/DXZLpYOxyVyy2WHmDh9jdb37Q2zPH5PmxN+hSjIs2MFVRT7OYHByN5W5AbAyT6UE5BAcUc2MvcYHVjL9Bzj2+U0p7nCUJJcOYgGy2gZJ9pdKFnZZ7vZ0mhl718tZkOof9MU9LXmUEd3pQAo3H+uW+Enai7XzBiyEYyoiTWmhbzMyvN2n1iSGqrWFfjCh0j6LIiZ1tfxi/11kP8PmckBYaTjijAlJvEleqoU5aO0dGmw/wv9rEi4w0rsevhVI43e5Dp20DpUrIExN5zssE206HGJ9DkmYdq7JLvGGgGE6D0/OxZD7bXOWwO1MhyuddZSPbDnH55R6XRmdPGkkMDHwGTY2dzW8jhx8ZAatncoP3rXIKzTeMC9PE/lH8stVuR+7BoiV4PougWnZ9bT2Zbs= ql2547@nyu.edu


In [ ]:
!ssh -T git@github.com

Hi ChelseaLiu0822! You've successfully authenticated, but GitHub does not provide shell access.


In [ ]:
!git clone --filter=blob:none --sparse  git@github.com:ChelseaLiu0822/LDRPM-Loan-Default-Risk-Prediction-Model.git
%cd LDRPM-Loan-Default-Risk-Prediction-Model
!git sparse-checkout add Data

fatal: destination path 'LDRPM-Loan-Default-Risk-Prediction-Model' already exists and is not an empty directory.
/content/LDRPM-Loan-Default-Risk-Prediction-Model
/content/LDRPM-Loan-Default-Risk-Prediction-Model/Data


## connect to GoogleDrive(Optional)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
%cd /content/drive/MyDrive/BigData/LDRPM-Loan-Default-Risk-Prediction-Model

/content


In [ ]:
!ls

OriginalData


## download data

In [ ]:
%cd Data

In [ ]:
!unzip OriginalData/parquet_files.zip

Archive:  OriginalData/parquet_files.zip
   creating: content/parquet_files/
   creating: content/parquet_files/test/
  inflating: content/parquet_files/test/test_credit_bureau_a_2_8.parquet  
  inflating: content/parquet_files/test/test_static_cb_0.parquet  
  inflating: content/parquet_files/test/test_debitcard_1.parquet  
  inflating: content/parquet_files/test/test_credit_bureau_a_2_7.parquet  
  inflating: content/parquet_files/test/test_base.parquet  
  inflating: content/parquet_files/test/test_applprev_1_2.parquet  
  inflating: content/parquet_files/test/test_credit_bureau_a_2_11.parquet  
  inflating: content/parquet_files/test/test_applprev_1_0.parquet  
  inflating: content/parquet_files/test/test_credit_bureau_a_2_9.parquet  
  inflating: content/parquet_files/test/test_tax_registry_c_1.parquet  
  inflating: content/parquet_files/test/test_credit_bureau_a_2_4.parquet  
  inflating: content/parquet_files/test/test_credit_bureau_a_1_2.parquet  
  inflating: content/parquet_

In [ ]:
!mv content/parquet_files OriginalData

In [ ]:
!rm -r content

In [ ]:
!pip install "dask[complete]"

In [ ]:
!python -m pip install "dask[dataframe]"   # Install requirements for dask dataframe

# Load Data

In [ ]:
import pandas as pd
import dask.dataframe as dd
import zipfile
import os
import itertools

In [ ]:
ROOT = 'OriginalData/'
TRAIN_DIR = ROOT+'train/'
TEST_DIR = ROOT+'test/'

## Load Features

In [ ]:
ROOT = 'OriginalData/'
TRAIN_DIR = ROOT+'parquet_files/train/'
TEST_DIR = ROOT+'parquet_files/test/'
base_store = {
    'base' : ['base'],
    'depth0' : ['static_0', 'static_cb_0'],
    'depth1' : ['applprev_1','other_1','tax_registry_a_1','tax_registry_b_1','tax_registry_c_1','credit_bureau_a_1','credit_bureau_b_1','deposit_1','person_1','debitcard_1'],
    'depth2' : ['applprev_2','person_2','credit_bureau_a_2','credit_bureau_b_2']
}

In [ ]:
feats_description = pd.read_parquet('OriginalData/feature_definitions_dtypes_tables.parquet')
feats_description

Variable  \
0              actualdpd_943P   
1     actualdpdtolerance_344P   
2        addres_district_368M   
3            addres_role_871L   
4             addres_zip_823M   
..                        ...   
460  totinstallast1m_4525188A   
461        twobodfilling_608L   
462                  type_25L   
463            typesuite_864L   
464           validfrom_1069D   

                                           Description    dtype  \
0    Days Past Due (DPD) of previous contract (actu...  Float64   
1                        DPD of client with tolerance.  Float64   
2                    District of the person's address.   String   
3                            Role of person's address.   String   
4                             Zip code of the address.   String   
..                                                 ...      ...   
460  Total amount of monthly instalments paid in th...  Float64   
461                       Type of application process.   String   
462                          Contact type of a person.   String   
463  Persons accompanying the client during the loa...   String   
464      Date since the client has an active campaign.     Date   

                                       tables  
0    [train_applprev_1_0, train_applprev_1_1]  
1        [train_static_0_0, train_static_0_1]  
2                            [train_person_2]  
3                            [train_person_2]  
4                            [train_person_2]  
..                                        ...  
460      [train_static_0_0, train_static_0_1]  
461      [train_static_0_0, train_static_0_1]  
462                          [train_person_1]  
463      [train_static_0_0, train_static_0_1]  
464      [train_static_0_0, train_static_0_1]  

[465 rows x 4 columns]

There's are 465 Variable & 436 Description

**P - Transform DPD (Days past due)**

Ex) DPD previous contract, DPD of client with tolerance

**M - Masking categories**

Ex) District of the person's address, Role of the person's address

**A - Transform amount**

Ex) Deposit Amount & Tax Deduction Amount

**D - Transform date**

Ex) Date of birth of the person, End date of deposit contract

**T - Unspecified Transform**

Ex) Type of income of the person, Month when the maximum Day Past Due (DPD) occurred for active contracts on credit bureau's records

**L - Unspecified Transform**

Ex) Type of application process,Interest rate for the active contracts

Please note that transformations within a group are denoted by a capital letter at the end of the predictor name (e.g., maxdbddpdtollast6m_4187119P). They hope that this will simplify the manipulation with predictors.

In [ ]:
# show all columns classified by depth and ended capital letter
classified_letters = ['P','M','A','D','T','L']
combinations = list(itertools.product(classified_letters, ['depth1','depth2']))
for pairs in combinations:
  print('variables ended with {0}\n'.format(pairs[0]))
  print('Variables in {0} tables'.format(pairs[1]))
  for index, row in feats_description[feats_description['Variable'].str.endswith(pairs[0])].iterrows():
    str_tables=','.join(row['tables'])
    if any(element in str_tables for element in base_store[pairs[1]]):
      print('\t'.join([row.Variable,row.Description,str_tables]))
  print('\n')

variables ended with P

Variables in depth1 tables
actualdpd_943P	Days Past Due (DPD) of previous contract (actual).	train_applprev_1_0,train_applprev_1_1
dpd_550P	The number of days past due for active loans where a guarantee has been provided.	train_credit_bureau_b_1
dpd_733P	Days past due (DPD) for guaranteed loans that were terminated according to credit bureau data.	train_credit_bureau_b_1
dpdmax_139P	Maximal days past due for active contract.	train_credit_bureau_a_1_0,train_credit_bureau_a_1_1,train_credit_bureau_a_1_2,train_credit_bureau_a_1_3
dpdmax_757P	Maximum days past due for a closed contract.	train_credit_bureau_a_1_0,train_credit_bureau_a_1_1,train_credit_bureau_a_1_2,train_credit_bureau_a_1_3
dpdmax_851P	Maximal past due days for active contracts in the credit bureau.	train_credit_bureau_b_1
maxdpdtolerance_577P	Maximum DPD with tolerance (on previous application/s).	train_applprev_1_0,train_applprev_1_1
pmtdaysoverdue_1135P	Number of days past due for existing contract

## Load Tables  


In [ ]:
for key, value in base_store.items():
  combinations = list(itertools.product(['train','test'], value))
  for pair in combinations:
    table_var = '_'.join(pair)
    file_path = locals()[pair[0].upper()+'_DIR']+table_var+'*.parquet'
    print(file_path)
    locals()[table_var] = dd.read_parquet(file_path)

OriginalData/parquet_files/train/train_base*.parquet
OriginalData/parquet_files/test/test_base*.parquet
OriginalData/parquet_files/train/train_static_0*.parquet
OriginalData/parquet_files/train/train_static_cb_0*.parquet
OriginalData/parquet_files/test/test_static_0*.parquet
OriginalData/parquet_files/test/test_static_cb_0*.parquet
OriginalData/parquet_files/train/train_applprev_1*.parquet
OriginalData/parquet_files/train/train_other_1*.parquet
OriginalData/parquet_files/train/train_tax_registry_a_1*.parquet
OriginalData/parquet_files/train/train_tax_registry_b_1*.parquet
OriginalData/parquet_files/train/train_tax_registry_c_1*.parquet
OriginalData/parquet_files/train/train_credit_bureau_a_1*.parquet
OriginalData/parquet_files/train/train_credit_bureau_b_1*.parquet
OriginalData/parquet_files/train/train_deposit_1*.parquet
OriginalData/parquet_files/train/train_person_1*.parquet
OriginalData/parquet_files/train/train_debitcard_1*.parquet
OriginalData/parquet_files/test/test_applprev_1*.

In [ ]:
# def transform_tables(df, is_base):
#   """
#   Function to transform a Dask DataFrame.
#   :param df: a Dask Dataframe
#       is_base<boolean>: is it a base file
#   :return: Dask DataFrame
#   """
#   df = dd.read_parquet(path)
#   if is_base:
#     df['date_decision'] = dd.to_datetime(df['date_decision'], format='%Y/%m/%d')
#   else:
#     for col in df.columns:
#       if col.endswith('M'):
#         # transform it to category dtype
#         df[col]=df[col].categorize()
#       elif col.endswith('D'):
#         df[col]=dd.to_datetime(df[col], format='%Y/%m/%d')

## Merge

In [ ]:
def describe_dask(df):
  """
  Function to compute descriptive statistics of a Dask DataFrame.
  :param df: Dask DataFrame
  :return: Pandas DataFrame with statistics
  """
  # Compute descriptive statistics
  desc = df.describe(percentiles=[0.25, 0.5, 0.75]).compute()

  # Compute total row count and missing values count
  row_count = len(df)
  missing = df.isnull().sum().compute()

  # Get data types
  data_types = df.dtypes

  # Creating a summary DataFrame
  summary = pd.DataFrame({
      'count': row_count,
      'missing': missing,
      'dtype': data_types,
      'min': desc.loc['min'],
      '25%': desc.loc['25%'],
      '50%': desc.loc['50%'],
      '75%': desc.loc['75%'],
      'max': desc.loc['max'],
      'mean': desc.loc['mean'],
      'std': desc.loc['std']
  })

  return summary.transpose()

# Example usage
# df = dd.read_csv('your_file.csv') # Replace with your file or DataFrame
# summary_df = describe_dask(df)
# print(summary_df)

base+depth0

In [ ]:
describe_dask(train_base)

MONTH   WEEK_NUM         case_id date_decision    target
count          1526659    1526659         1526659       1526659   1526659
missing              0          0               0             0         0
dtype            int64      int64           int64        object     int64
min           201901.0        0.0             0.0           NaN       0.0
25%           201906.0       23.0        766197.5           NaN       0.0
50%           201910.0       40.0       1357358.0           NaN       0.0
75%           202001.0       55.0       1739022.5           NaN       0.0
max           202010.0       91.0       2703454.0           NaN       1.0
mean     201936.287982  40.769036  1286076.571739           NaN  0.031437
std          44.735975  23.797981   718946.592285           NaN  0.174496

**notice that mean value of target **

In [ ]:
len(train_static_0.columns)

168

In [ ]:
len(train_static_cb_0.columns)

53

In [ ]:
train_static_cb_0.case_id.unique().compute()

0              357
1              381
2              388
3              405
4              409
            ...   
1500471    2703450
1500472    2703451
1500473    2703452
1500474    2703453
1500475    2703454
Name: case_id, Length: 1500476, dtype: int64

In [ ]:
train_depth0 = train_static_0.merge(train_static_cb_0, on='case_id', how='left')

In [ ]:
len(train_depth0.case_id)

1526659

In [ ]:
train_base_depth0 = train_base.merge(train_depth0, on='case_id', how='left')

In [ ]:
len(train_base_depth0.columns)

224

### 处理train_base_depth0里的缺失值

In [ ]:
train_depth0_NAs = train_static_cb_0[train_static_cb_0.isna().any(axis=1)]
col_contain_NA = train_static_cb_0.isnull().any()

In [ ]:
cols_with_missing = ['case_id']+col_contain_NA[col_contain_NA].index.compute().tolist()
static_with_missing_cols = train_depth0_NAs[cols_with_missing]

In [ ]:
static_with_missing_cols.head()

case_id assignmentdate_238D assignmentdate_4527235D  \
0      357                None                    None   
1      381                None                    None   
2      388                None                    None   
3      405                None                    None   
4      409                None                    None   

  assignmentdate_4955616D birthdate_574D  contractssum_5085716L  \
0                    None     1988-04-01                    NaN   
1                    None     1973-11-01                    NaN   
2                    None     1989-04-01                    NaN   
3                    None     1974-03-01                    NaN   
4                    None     1993-06-01                    NaN   

  dateofbirth_337D dateofbirth_342D  days120_123L  days180_256L  ...  \
0             None             None           NaN           NaN  ...   
1             None             None           NaN           NaN  ...   
2       1989-04-01             None           6.0           8.0  ...   
3       1974-03-01             None           0.0           0.0  ...   
4       1993-06-01             None           2.0           3.0  ...   

   pmtscount_423L  pmtssum_45A  requesttype_4525192L  responsedate_1012D  \
0             6.0    6301.4000                  None          2019-01-25   
1             6.0    4019.6000                  None          2019-01-25   
2             6.0   14548.0000                  None          2019-01-28   
3             6.0   10498.2400                  None          2019-01-21   
4             7.0    6344.8804                  None          2019-01-21   

   responsedate_4527233D  responsedate_4917613D  riskassesment_302T  \
0                   None                   None                None   
1                   None                   None                None   
2                   None                   None                None   
3                   None                   None                None   
4                   None                   None                None   

   riskassesment_940T  secondquarter_766L  thirdquarter_1082L  
0                 NaN                 NaN                 NaN  
1                 NaN                 NaN                 NaN  
2                 NaN                 3.0                 5.0  
3                 NaN                 2.0                 0.0  
4                 NaN                 0.0                 4.0  

[5 rows x 48 columns]

In [ ]:
for col in cols_with_missing:
    tar = feats_description[feats_description['Variable'] == col]

    # 确保找到匹配的列
    if not tar.empty:
        # 如果找到了匹配的行，打印相关信息
        print('\t'.join(tar.iloc[0].astype(str)))
    else:
        # 如果没有找到匹配的行，打印说明
        print(f"No description found for column: {col}")

No description found for column: case_id
assignmentdate_238D	Tax authority data - date of assignment.	Date	['train_static_cb_0']
assignmentdate_4527235D	Tax authority data - Date of assignment.	Date	['train_static_cb_0']
assignmentdate_4955616D	Tax authority assignment date.	Date	['train_static_cb_0']
birthdate_574D	Client's date of birth (credit bureau data).	Date	['train_static_cb_0']
contractssum_5085716L	Total sum of values of contracts retrieved from external credit bureau.	Float64	['train_static_cb_0']
dateofbirth_337D	Client's date of birth.	Date	['train_static_cb_0']
dateofbirth_342D	Client's date of birth.	Date	['train_static_cb_0']
days120_123L	Number of credit bureau queries for the last 120 days.	Float64	['train_static_cb_0']
days180_256L	Number of credit bureau queries for last 180 days.	Float64	['train_static_cb_0']
days30_165L	Number of credit bureau queries for the last 30 days.	Float64	['train_static_cb_0']
days360_512L	Number of Credit Bureau queries for last 360 days

To save memory:  
- categorize variables
- use float32

In [ ]:

for col in train_base_depth0.columns:
  if col.endswith('M'):
    train_base_depth0[col] = train_base_depth0[col].astype('category')
  elif train_base_depth0[col].dtype == 'float64':
    train_base_depth0[col] = train_base_depth0[col].astype('float32')

To get feature importance

In [ ]:
# !pip uninstall dask_xgboost

Found existing installation: dask-xgboost 0.2.0
Uninstalling dask-xgboost-0.2.0:
  Would remove:
    /usr/local/lib/python3.10/dist-packages/dask_xgboost-0.2.0.dist-info/*
    /usr/local/lib/python3.10/dist-packages/dask_xgboost/*
Proceed (Y/n)? Y
  Successfully uninstalled dask-xgboost-0.2.0


In [ ]:
# !pip install --upgrade xgboost

In [ ]:
import xgboost
print(xgboost.__version__)

2.0.3


**确定更新到了1.0版本以上**  
如果是在colab里请更新后重新启动运行时

In [ ]:
from dask.distributed import Client
import dask.dataframe as dd
import xgboost as xgb

client = Client()

X = train_base_depth0.drop('target', axis=1)
y = train_base_depth0['target']

dtrain = xgb.dask.DaskDMatrix(client, X, y)

params = {
    'objective': 'binary:logistic',
    'eval_metric': 'logloss'
}

output = xgb.dask.train(client, params, dtrain, num_boost_round=10)

bst = output['booster']

importance = bst.get_score(importance_type='weight')

for feature, score in importance.items():
    print(f'Feature: {feature}, Score: {score}')


INFO:distributed.http.proxy:To route to workers diagnostics web server please install jupyter-server-proxy: python -m pip install jupyter-server-proxy
INFO:distributed.scheduler:State start
INFO:distributed.diskutils:Found stale lock file and directory '/tmp/dask-scratch-space/scheduler-2vg666up', purging
INFO:distributed.diskutils:Found stale lock file and directory '/tmp/dask-scratch-space/scheduler-2ae34xj8', purging
INFO:distributed.diskutils:Found stale lock file and directory '/tmp/dask-scratch-space/scheduler-96uitlzk', purging
INFO:distributed.diskutils:Found stale lock file and directory '/tmp/dask-scratch-space/scheduler-r_zahs_3', purging
INFO:distributed.diskutils:Found stale lock file and directory '/tmp/dask-scratch-space/scheduler-wiqq0bbh', purging
INFO:distributed.scheduler:  Scheduler at:     tcp://127.0.0.1:42815
INFO:distributed.scheduler:  dashboard at:  http://127.0.0.1:8787/status
INFO:distributed.nanny:        Start Nanny at: 'tcp://127.0.0.1:39375'
INFO:distrib

AssertionError: error

riskassesment_302T	Estimated probability that the client will default on their credit obligation within the next year.	String	['train_static_cb_0']

riskassesment_940T	Estimate of client's creditworthiness.	Float64	['train_static_cb_0']
